Author "Return Type Notation MVP: Call for Testing!" (#1408)
rust-lang/blog.rust-lang.org: posts/inside-rust/2024-09-23-rtn-call-for-testing.md

author: Michael Goulet
team: The Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
---

The async working group is excited to announce that [RFC 3654] return type notation (RTN) is ready for testing. In this post, we want to briefly describe the feature and announce a call for testing on nightly Rust.

## The backstory

Rust 1.75 [stabilized](https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html) async fn in traits (AFIT) and return-position impl Trait in traits (RPITIT). These desugar to anonymous generic associated types (GATs). However, unlike GATs, users of these types cannot use `where` clauses to further restrict these return types. This is known as the ["send bound"](https://smallcultfollowing.com/babysteps/blog/2023/02/01/async-trait-send-bounds-part-1-intro/) problem, since it often affects `Send` bounds on futures in the async ecosystem.

### An example

Consider a trait `Foo` with a `method` that returns a type of `impl Future<Output = ()>`. We want to write a function that calls `method` and spawns the future on another thread:

```rust
fn spawn<T>(f: impl Future<Output = T> + Send + 'static) {}

trait Foo {

>     fn method() -> impl Future<Output = ()>; // <-- RPITIT.

Reviewer: I wonder if it's worth spelling out RPITIT anywhere?

Author: It's spelled out in the first sentence of `The backstory` section

Reviewer: Wow, reading comprehension fail! Sorry about that.

}

fn needs_sendable_future<T: Foo>()
where
    // How do we further restrict `T::method()` to be `Send + 'static`?
{
    spawn(T::method());
    //~^ ERROR: `impl Future<Output = ()>` is not `Send`!
}
```

Specifically, we may not want to restrict the *declaration* of `Foo`, since changing it in the declaration would restrict *all* implementations of `Foo`.

```rust
trait Foo {
    fn method() -> impl Future<Output = ()> + Send + 'static;
    //                                      ~~~~~~~~~~~~~~~~
    //                                      Not what we want.
}
```

> So, on stable Rust, we have no way of expressing this restriction when using AFIT or RPITIT. In contrast, we can express this today if we were to use a GAT directly:

Reviewer: ```suggestion
So, on stable Rust, we have no way of expressing this restriction when using async functions in traits or RPITIT. In contrast, we can express this today if we were to use a generic associated type (GAT) directly:
```

Author: Why did you change AFIT but not RPITIT? All of these acronyms are spelled out in the backstory section; we could either consistently use acronyms or the full names, but this seems a bit inconsistent... Your edit is literally doing all three of:
* Only full name
* Only acronym
* Both full name and acronym

Reviewer: It's inconsistent (and wrong because all these are spelled out in the backstory section), but my thinking was that I'd only seen AFIT used once so we might as well spell it out and not bother with the acronym, RPITIT had already been used, and GAT was used later.

Anyway, this suggestion doesn't make sense so I withdraw it!

```rust
trait Foo {
    type MethodFuture: Future<Output = ()>;
    fn method() -> Self::MethodFuture;
}

fn needs_sendable_future<T: Foo>()
where
    // We can restrict this to only implementors of `Foo`
    // whose `MethodFuture` is `Send + 'static`, so we can
    // call `spawn` below:
    T::MethodFuture: Send + 'static
{
    spawn(T::method());
}
```

However, using GATs means that implementors of `Foo` have to write out the return type explicitly, `type MethodFuture = ...`, which doesn't ([yet](https://github.com/rust-lang/rust/pull/120700)) work if we have an anonymous, unnameable `Future` type!

## The solution

> In [RFC 3654] we introduced return type notation (RTN). This will allow us to write `where` clause bounds that restrict the return types of functions and methods that use async fn in traits (AFIT) and return-position impl Trait in traits (RPITIT). Extending the example above, RTN lets us write:

Reviewer: I like that you spell out all the acronyms here but it'd be good to pull that earlier in the post.

Author: They are earlier in the post :)

Reviewer: 🤦

```rust
fn needs_sendable_future<T: Foo>()
where
    T::method(..): Send + 'static // ✨
{
    spawn(T::method());
    //~^ Works!
}
```

## Restrictions

Currently, RTN is only allowed for trait associated functions and methods with lifetime generics (not const or type generics) that use:

* async fn in traits (AFIT)
* return-position impl Trait in traits (RPITIT) where the impl Trait is the outermost return type, i.e. `-> impl Trait`, but not `-> Box<impl Trait>`

These restrictions are described in further detail in [RFC 3654].

## How do I help?

> We'd love for you to test out this new feature!

Author: ```suggestion
We'd love for you to test out this feature on the latest Rust nightly compiler[^nightly].

[^nightly]: Make sure to run `rustup update nightly` or update your compiler with your system package manager, since the feature is very new.
```

Author: Perhaps?

Reviewer: By "system package manager" do you mean `apt` or `yum` or whatever? Do those usually have nightly in them?

I like this change overall because it's more precise. Before it could be interpreted as if the feature were already on stable.

Author: > By "system package manager" do you mean apt or yum or whatever? Do those usually have nightly in them?

Well I'm just trying to cover "people who dont use rustup" here.

Reviewer: I like that! It reminds me of "or wherever you get your podcasts."

---
layout: post

> title: "Return Type Notation MVP: Call for Testing!"

Reviewer 2: ```suggestion
title: "Return type notation MVP: Call for testing!"
```

We've been rather inconsistent about this, but the blog mostly uses sentence-case titles, which has been the trend on the web.  (We might call the capitalization after the colon "piecewise sentence-case".)

Author: Sure, I don't care either way.

## The solution

In [RFC 3654] we introduced return type notation (RTN). This will allow us to write `where` clause bounds that restrict the return types of functions and methods that use async fn in traits (AFIT) and return-position impl Trait in traits (RPITIT). Extending the example above, RTN lets us write:

```rust
fn needs_sendable_future<T: Foo>()
where
    T::method(..): Send + 'static // ✨
{
    spawn(T::method());
    //~^ Works!
}
```

## Restrictions

Currently, RTN is only allowed for trait associated functions and methods with lifetime generics (not const or type generics) that use:

* async fn in traits (AFIT)
* return-position impl Trait in traits (RPITIT) where the impl Trait is the outermost return type, i.e. `-> impl Trait`, but not `-> Box<impl Trait>`

These restrictions are described in further detail in [RFC 3654].

## How do I help?

We'd love for you to test out this feature on the latest Rust nightly compiler[^nightly].

> [^nightly]: Make sure to run `rustup update nightly` or however manage your Rust releases, since the feature is very new and is still unstable!

Author: ```suggestion
[^nightly]: Make sure to run `rustup update nightly` or however you manage your Rust releases, since the feature is very new and is still unstable!
```
